arrow/src/main/java/org/apache/iceberg/arrow/ArrowSchemaUtil.java

```diff
@@ -51,94 +54,133 @@ private ArrowSchemaUtil() {}
   public static Schema convert(final org.apache.iceberg.Schema schema) {
     ImmutableList.Builder<Field> fields = ImmutableList.builder();
 
-    for (NestedField f : schema.columns()) {
-      fields.add(convert(f));
+    for (NestedField field : schema.columns()) {
+      fields.add(TypeUtil.visit(field.type(), new IcebergToArrowTypeConverter(field)));
```

Comment:
Rather than operate field-by-field like this, it would make more sense for the schema visitor to take the passed in schema and convert to perform the complete conversion that way.  This approach is mixing the patterns, but the visitor pattern covers this as well.

That would make this a straightforward:

```java
return TypeUtil.visit(schema, new IcebertToArrowTypeConvertor(schema));
```